pkg/blobstore/spanner_blob_access.go

```diff
@@ -0,0 +1,975 @@
+		log.Printf("Can't create GCS client: %v", err)
+		return nil, err
+	}
+	gcsBucket := storageClient.Bucket(gcsBucketName)
+
+	// If the GCS bucket doesn't exist, create it.
+	_, err = gcsBucket.Attrs(ctx)
+	if err == storage.ErrBucketNotExist {
+		err = createGCSBucket(ctx, gcsBucket, databaseName, daysToLive)
+		if err != nil {
+			// We could have raced with another pod.  Check if the bucket exists.
+			_, xerr := gcsBucket.Attrs(ctx)
+			if xerr == storage.ErrBucketNotExist {
+				// Bucket still doesn't exist.
+				spannerClient.Close()
+				storageClient.Close()
+				log.Printf("Can't create GCS bucket: %v", err)
+				return nil, err
+			}
+		}
+	} else if err != nil {
+		spannerClient.Close()
+		storageClient.Close()
+		log.Printf("Can't access GCS bucket: %v", err)
+		return nil, err
+	} else {
+		// Check if we need to update the TTL.
+		days, err := getGCSTTL(ctx, gcsBucket)
+		if err != nil {
+			log.Printf("Can't determine GCS TTL: %v", err)
+		}
+		if days != daysToLive {
+			err = updateGCSDeletionPolicy(ctx, gcsBucket, daysToLive)
+			if err != nil {
+				log.Printf("Can't update GCS TTL: %v", err)
+			} else {
+				log.Printf("GCS TTL changed from %d to %d days", days, daysToLive)
+			}
+		}
+	}
+
+	log.Printf("NewSpannerBlobAccess type %#+v", storageType)
+
+	// FindMissing takes care of updaing the reference time on CAS objects, but we'd like to update
+	// AC objects when they're read, to simulate an LRU cache.  Doing this one at a time is inefficient.
+	var refCh chan string
+	if storageType == pb.StorageType_ACTION_CACHE {
+		refCh = make(chan string, maxRefBulkSz)
+	}
+
+	ba := &spannerBlobAccess{
+		spannerClient: spannerClient,
+		gcsBucket:     gcsBucket,
+
+		readBufferFactory: readBufferFactory,
+		storageType:       storageType,
+		daysToLive:        daysToLive,
+		expirationAge:     expirationAge,
+		refUpdateThresh:   refUpdateThresh,
+		refChan:           refCh,
+	}
+	if refCh != nil {
+		go ba.bulkUpdate(refCh)
+	}
+	return ba, nil
+}
+
+func (ba *spannerBlobAccess) GetCapabilities(ctx context.Context, instanceName digest.InstanceName) (*remoteexecution.ServerCapabilities, error) {
+	if ba.storageType == pb.StorageType_ACTION_CACHE {
+		return &remoteexecution.ServerCapabilities{
+			CacheCapabilities: &remoteexecution.CacheCapabilities{
+				ActionCacheUpdateCapabilities: &remoteexecution.ActionCacheUpdateCapabilities{
+					UpdateEnabled: true,
+				},
+				SymlinkAbsolutePathStrategy: remoteexecution.SymlinkAbsolutePathStrategy_ALLOWED,
+			},
+		}, nil
+	} else if ba.storageType == pb.StorageType_CASTORE {
+		return &remoteexecution.ServerCapabilities{
+			CacheCapabilities: &remoteexecution.CacheCapabilities{
+				DigestFunctions: digest.SupportedDigestFunctions,
+			},
+		}, nil
+	} else {
+		// Code in pkg/blobaccess/configuration/new_blob_access.go should prevent this, but
+		// since that lives in a separate place, let's be sure.
+		panic("Invalid Spanner storage Type configured")
+	}
+}
```

Comment:
Any reason this can't leverage `capabilities.Provider`?

Reply:
I was getting weird results in the capabilities message, so I rolled my own.  I'll try to get you an example.

Reply:
The anomaly occurs regardless of whether I use capabilities.Provider or roll my own code:

Capabilities for gs.com received:
cache_capabilities:{digest_functions:MD5  digest_functions:SHA1  digest_functions:SHA256  digest_functions:8  digest_functions:SHA384  digest_functions:SHA512  action_cache_update_capabilities:{}  symlink_absolute_path_strategy:ALLOWED}  deprecated_api_version:{major:2}  low_api_version:{major:2}  high_api_version:{major:2  minor:3}

Capabilities for gs.com/unreviewed received:
cache_capabilities:{digest_functions:MD5  digest_functions:SHA1  digest_functions:SHA256  digest_functions:8  digest_functions:SHA384  digest_functions:SHA512  action_cache_update_capabilities:{update_enabled:true}  symlink_absolute_path_strategy:ALLOWED}  execution_capabilities:{digest_function:SHA256  exec_enabled:true  execution_priority_capabilities:{priorities:{min_priority:-2147483648  max_priority:2147483647}}  5:"\x03\x02\x01\x08\x05\x06"}  deprecated_api_version:{major:2}  low_api_version:{major:2}  high_api_version:{major:2  minor:3}

Notice the 5:"\x03\x02\x01\x08\x05\x06" that occurs in the execution_capabilities in the second case.

Reply:
Whatever tool you used to dump these message, make sure it uses an up-to-date version of remote_execution.proto. Once you do, you'll notice that field 5 in ExecutionCapabilities is actually `digest_functions`:

https://github.com/bazelbuild/remote-apis/commit/64cc5e9e422c93e1d7f0545a146fd84fcc0e8b47

And what Buildbarn does here is expected. Namely, it reports the list of digest functions that are supported by the scheduler.

Reply:
Yep, rebuilding the tool prints the hash functions.  I hadn't realized the remote APIs has changed that much.